### PR TITLE
moderator actions buildout

### PR DESCRIPTION
**Pull request by b0b3rt**
image

Builds out the initial infrastructure for automoderation, centered on the concept of "moderator actions" which can be applied to a user.  Starts with a couple basic cases (comment quality warnings) which don't yet have any user-facing effects, but flag a user with `needsReview` (so that we can see what things would look like to tune the rules as necessary, prior to actually rolling out anything user-facing):
- if 2 or more of a user's last 5 comments are net-downvoted
- if the average karma of a user's last 20 comments is under 1.5

Also adds functionality for post and comment rate-limiting based on another moderator action type, but nothing yet triggers this action.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203155014071287) by [Unito](https://www.unito.io)

### CODE REVIEW

**Review by Raemon — APPROVED** [review]

File: packages/lesswrong/server/callbacks/rateLimits.ts [hunk 1]
```diff
@@ -48,4 +62,40 @@ async function enforcePostRateLimit (user: DbUser) {
...
   }
+
+  const moderatorRateLimit = await ModeratorActions.findOne({
+    userId: user._id,
+    type: RATE_LIMIT_ONE_PER_DAY,
+    endedAt: null
+  });
+
+  if (numberOfPostsInPast24Hours > 0 && moderatorRateLimit) {
+    throw new Error(`You have been rate limited to 1 post per day.`);
```
Comment (darkruby501) [inline packages/lesswrong/server/callbacks/rateLimits.ts:73]: If applies to both posts and comments, wording should reflect this

File: packages/lesswrong/server/callbacks/sunshineCallbackUtils.tsx [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
+import { createMutator } from "../vulcan-lib";
+import { COMMENT_LOW_QUALITY_WARNING, COMMENT_MEDIOCRE_QUALITY_WARNING, isActionActive } from "../../lib/collections/moderatorActions/schema";
+import { forumTypeSetting } from "../../lib/instanceSettings";
 
 /** This function contains all logic for determining whether a given user needs review in the moderation sidebar.
  * 
  * It's important this this only be be added to async callbacks on posts and comments, so that postCount and commentCount have time to update first
  */
-export async function triggerReviewIfNeeded(userId: string) {
+export async function triggerReviewIfNeeded(userId: string, newModeAction?: true) {
```
Comment (darkruby501) [inline packages/lesswrong/server/callbacks/sunshineCallbackUtils.tsx]: @Raemon I don't like this variable name. Forces me to look inside the function to know what it does. Seems it's function is something like "override", in which case call it that.

(Tagging you 'cause Robert says you thought it should be this way.)

File: packages/lesswrong/server/callbacks/sunshineCallbackUtils.tsx [hunk 3]
```diff
@@ -23,7 +28,71 @@ export async function triggerReviewIfNeeded(userId: string) {
...
+}
+
+function areLowQualityComments(comments: DbComment[]) {
+  const threshold = 2;
+  const downvotedCommentCount = comments.filter(isNetDownvoted).length;
+
+  return downvotedCommentCount >= threshold;
+}
+
+function areMediocreQualityComments(comments: DbComment[]) {
```
Comment (darkruby501) [inline packages/lesswrong/server/callbacks/sunshineCallbackUtils.tsx:52]: Might prefer a name like "commentsHaveLowAverageKarma" or something that is transparent about happens

Comment (darkruby501) [reply packages/lesswrong/server/callbacks/sunshineCallbackUtils.tsx:52]: (not certain about this one)

**Pull request merged**